group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroupTest.java

```diff
@@ -0,0 +1,500 @@
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(fooTopicId, 7));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(fooTopicId, 8));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(fooTopicId, 9));
+    }
+
+    @Test
+    public void testDeletingMemberRemovesPartitionEpoch() {
+        Uuid fooTopicId = Uuid.randomUuid();
+        Uuid barTopicId = Uuid.randomUuid();
+        Uuid zarTopicId = Uuid.randomUuid();
+
+        ConsumerGroup consumerGroup = createConsumerGroup("foo");
+        ConsumerGroupMember member;
+
+        member = new ConsumerGroupMember.Builder("member")
+            .setMemberEpoch(10)
+            .setAssignedPartitions(mkAssignment(
+                mkTopicAssignment(fooTopicId, 1, 2, 3)))
+            .setPartitionsPendingRevocation(mkAssignment(
+                mkTopicAssignment(barTopicId, 4, 5, 6)))
+            .setPartitionsPendingAssignment(mkAssignment(
+                mkTopicAssignment(zarTopicId, 7, 8, 9)))
+            .build();
+
+        consumerGroup.updateMember(member);
+
+        assertEquals(10, consumerGroup.currentPartitionEpoch(fooTopicId, 1));
+        assertEquals(10, consumerGroup.currentPartitionEpoch(fooTopicId, 2));
+        assertEquals(10, consumerGroup.currentPartitionEpoch(fooTopicId, 3));
+        assertEquals(10, consumerGroup.currentPartitionEpoch(barTopicId, 4));
+        assertEquals(10, consumerGroup.currentPartitionEpoch(barTopicId, 5));
+        assertEquals(10, consumerGroup.currentPartitionEpoch(barTopicId, 6));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(zarTopicId, 7));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(zarTopicId, 8));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(zarTopicId, 9));
+
+        consumerGroup.removeMember(member.memberId());
+
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(barTopicId, 1));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(barTopicId, 2));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(barTopicId, 3));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(zarTopicId, 4));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(zarTopicId, 5));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(zarTopicId, 6));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(fooTopicId, 7));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(fooTopicId, 8));
+        assertEquals(-1, consumerGroup.currentPartitionEpoch(fooTopicId, 9));
+    }
+
+    @Test
+    public void testGroupState() {
+        ConsumerGroup consumerGroup = createConsumerGroup("foo");
+        assertEquals(ConsumerGroup.ConsumerGroupState.EMPTY, consumerGroup.state());
+
+        ConsumerGroupMember member1 = new ConsumerGroupMember.Builder("member1")
+            .setMemberEpoch(1)
+            .setPreviousMemberEpoch(0)
+            .setNextMemberEpoch(1)
```

Comment:
Is there ever a time the next member epoch is not the same as the member epoch?

Reply:
i am curious as well. i notice 3 places
```
CurrentAssignmentBuilder#maybeTransitionFromAssigningToAssigningOrStable
CurrentAssignmentBuilder#transitionToNewTargetAssignmentState
CurrentAssignmentBuilder#maybeTransitionFromRevokingToAssigningOrStable
```
that both set member epoch and next member epoch to `targetAssignmentEpoch`

Reply:
That's right. There is only one case where it is not. When a member must revoke partitions, it stays in its current epoch so member epoch is different from the next epoch in this case.

The rational of keeping track of the next epoch here is to basically prevent recomputing the state while the member is in revoking state. Without it, we would have to recompute it on every heartbeat.

Reply:
Is nextMemberEpoch a bit confusing here? It seems more like a target epoch.

Reply:
Yeah, I agree. Let me use `targetMemberEpoch`.